lib/icinga/apiactions.cpp

```diff
@@ -681,6 +681,15 @@ Dictionary::Ptr ApiActions::ExecuteCommand(const ConfigObject::Ptr& object, cons
 	if (!endpointPtr)
 		return ApiActions::CreateResult(404, "Can't find a valid endpoint for '" + resolved_endpoint + "'.");
 
+    /* Check if the endpoint zone can access the checkable */
+    Zone::Ptr endpointZone = endpointPtr->GetZone();
+    if (!endpointZone->CanAccessObject(checkable)) {
```

Comment:
This check prevents executing commands for checkables with `command_endpoint` in a child zone on that agent. So for example with `agent-1!hostname` having `zone = "master"` and `command_endpoint = "agent-1"` where that agent-1 is a direct child of the master zone, this results in the following error:

```console
$ curl -sSku root:icinga https://localhost:5665/v1/actions/execute-command --json '{"type": "Service", "service": "agent-1!hostname", "ttl": 15, "macros": {"command_endpoint": "agent-1"}, "command": "hostname", "command_type": "CheckCommand", "pretty": true}'
{
    "results": [
        {
            "code": 409,
            "status": "Zone 'agent-1' cannot access checkable 'agent-1!hostname'."
        }
    ]
}
```

So the check should probably be whether the parent zone of `endpointZone` (if it exists, otherwise that zone itself) can access `checkable` to allow for that situation.

Reply:
Hi @julianbrost, since I'm not able to reproduce the problem may I ask you to confirm if the following configuration is the one that you used for the test or if you can provide me the correct one? Thank you

master zones.conf
```
object Endpoint "master" {
}

object Zone "master" {
    endpoints = [ "master", "agent-1" ]
}

object Endpoint "agent-1" {
    host = "127.0.0.1"
    port = "5662"
}
```

master.conf
```
object Host "master" {
    address = "127.0.0.1"
    check_command = "hostalive"
}

object Host "agent-1" {
    address = "127.0.0.1"
    check_command = "hostalive"
    zone = "master"
    enable_active_checks = false
}

object Service "hostname" {
    host_name = "agent-1"
    check_command = "ping"
}
```

agent-1 zones.conf
```
object Endpoint "master" {
    host = "127.0.0.1"
    port = "5665"
}

object Zone "master" {
    endpoints = [ "master", "agent-1" ]
}

object Endpoint "agent-1" {
}
```
agent-1.conf
```
object Host "agent-1" {
    address = "127.0.0.1"
    check_command = "hostalive"
    zone = "master"
    enable_active_checks = false
}

object Service "hostname" {
    host_name = "agent-1"
    check_command = "ping"
}
```

Reply:
> object Zone "master" {
>     endpoints = [ "master", "agent-1" ]
> }

The agent-1 Endpoint object has to be in its own zone. That config above would be for an HA master zone.

So more like this:

master `zones.conf`:
```
object Endpoint "master" {
}

object Zone "master" {
    endpoints = [ "master" ]
}

object Endpoint "agent-1" {
    host = "127.0.0.1"
    port = "5662"
}

object Zone "agent-1" {
    parent = "master"
    endpoints = [ "agent-1" ]
}
```

master `zones.d/master/agent-1.conf`:
```
object Host "agent-1" {
    address = "127.0.0.1"
    check_command = "hostalive"
    enable_active_checks = false
}

object Service "hostname" {
    host_name = "agent-1"
    check_command = "ping"
    command_endpoint = "agent-1"
}
```

agent-1 `zones.conf`:
```
object Endpoint "master" {
    host = "127.0.0.1"
    port = "5665"
}

object Zone "master" {
    endpoints = [ "master" ]
}

object Endpoint "agent-1" {
}

object Zone "agent-1" {
    parent = "master"
    endpoints = [ "agent-1" ]
}
```

And no additional Host/Service objects for `agent-1!hostname` on the agent, due to `command_endpoint = "agent-1"` they don't have to be present there.

Reply:
I did some testing and it seems that using the parent zone with the following configuration:
```
       m
      / \
     /   \
    /     \
  s-1     s-2  
```
it's possible to execute a command with checkable `s-1` in the `s-2` endpoint because the master zone can access `s-2`

I suggest checking if the endpointZone can access the checkable only when the endpoint where I want to execute the command is different than the command endpoint of the checkable, so with a check like the following one

```
    Zone::Ptr endpointZone = endpointPtr->GetZone();
    Endpoint::Ptr commandEndpoint = checkable->GetCommandEndpoint();
    if (endpointPtr != commandEndpoint && !endpointZone->CanAccessObject(checkable)) {
```

What do you think?

Reply:
> it's possible to execute a command with checkable `s-1` in the `s-2` endpoint because the master zone can access `s-2`

That doesn't come as a surprise so me, that's a simpler scenario of one what I described in https://github.com/Icinga/icinga2/pull/8627#issuecomment-780581632:

> 1. Restricting the `endpoint` parameter of the `/v1/actions/execute-command` API to those values, where we could currently execute and retrieve the result. That is the object's zone, all its parent zones and any direct child of a aforementioned zone. In this example, for an object in zone `s-2-1`, checks could be executed on `m`, `s-2`, `s-2-1`, `a-2` and `s-1` (but **not** `a-1`). For me this feels like a quite arbitrary restriction from a user's perspective.

Last sentence is still true, so yes, that's a bit strange. But it's not a huge problem, `s-2` can't impersonate `s-1`, it can only update an execution on `s-1` if that execution was sent to `s-2`, which is somewhat of a strange input in the first place.

> ```
>     Zone::Ptr endpointZone = endpointPtr->GetZone();
>     Endpoint::Ptr commandEndpoint = checkable->GetCommandEndpoint();
>     if (endpointPtr != commandEndpoint && !endpointZone->CanAccessObject(checkable)) {
> ```

So with this change compared to the current state of the PR, it would additionally accept the command in the very specific `command_endpoint` case. Sounds reasonable, so if that works, go ahead.